### ISSUE TITLE
RFC: reporting module

### ISSUE THREAD

**Issue by Dziolas** [inveniosoftware/invenio#2895]
The idea for a reporting module come from a couple of use cases that I have for SCOAP3. The main principle would be to have plugin based module for periodical/on event generation of "reports". Reports could be:
- stored (for a future reference) or send to specified email addresses,
- in a different formats: email, PDF, CSV etc.,
- they could be triggered periodically (every day/week etc) or on event (faild bibcheck etc.)

In general it could be simillar to bibcheck/checker but should not change data and scheduling should be done per plugin.

Use cases that I have for SCOAP3:
- reports of status (number of records, and some other queried/dynamically generated statistics) of repository sent to SCOAP3 partners,
- report about ingestion and compliance checks (we check metadata, PDF and publishers website for certain things - this are not only simple checks) created when compliance check fails on ingestion of record to the system or periodically as summary
- ad hoc reports in CSV format for some specific queries metadata.

Any comments/ideas:question:

**Comment by lnielsen** [comment]
Yes, it likely goes hand in hand with data collection. Usually I think you want to run a query and store the number. I had something in mind for the data collection.

**Comment by PXke** [comment]
Just to know, Do you plan to base it on the statistics module or make it  fully independent  ?

**Comment by Dziolas** [comment]
@lnielsen I'm not sure if i know what are the data collections? What do you mean by them?
@PXke I was thinking about something independent that could communicate with other modules to get and receive data (except from plugins) - it might be handy to just use some function from reporting module to generate a pdf with results of some task and then automatically send it email address (report would still be registered in the system so you could see it later).

**Comment by lnielsen** [comment]
Data collection I think is what you call generate data.

**Comment by tiborsimko** [comment]
CC @aw-bib and @shesselbach who have extensive reporting needs in the HGF use case.  These are indeed of "statistics" nature.

**Comment by tiborsimko** [comment]
@Dziolas I like the idea of a "reports" module offering the above capabilities.  One more use case would be to link it to the future "audit store" weekly reports. Concerning "heavy computations", how would you define boundaries what is a gathering-data report and what as a generating-data report?  This could largely depend on the system size; what is light for 100M record system could be heavy for 1000M record system.  Perhaps you can treat all reporting needs as equal citizens and off-load all generations to independent Celery tasks.

**Comment by aw-bib** [comment]
I may add that one of the main points of our statistics class is _not_ to produce only numbers, but also allow for a link to the publications counted. Something like online publication lists. User feedback tells us that this is indeed important. ("Are all _my_ publications included in the bean counting?")

Besides we offer functionality to produce the final publications lists via BibTeX -> BibLaTeX -> Biber -> LuaLaTeX -> PDF. Now this can hardly be done interactively. (We mused about some robot ramping approach similar to arXiv. Nothing done yet.)

Finally: some of our reports are of a nature that in Invenio 1.x there is no query so we need to loop over (large) `intbitsets` (Count for grant ID $0 if and only if $x==0 if you have several grants assigned. e.g.) I'm not really sure that this can change in general. Depends a bit on the stats.

**Comment by tiborsimko** [comment]
> Count for grant ID $0 if and only if $x==0 if you have several grants assigned. e.g.) I'm not really sure that this can change in general.

Without knowing about the background of such a use case, if you have complex queries depending on the values of several fields/subfields, perhaps you can write a new dedicated tokeniser that would produce a new index to help with these queries?  Kind of like `journal` index, or composed `author:affiliation` one?

**Comment by aw-bib** [comment]
@tiborsimko I understood that this is only available for invenio >(>?) 1.1, but I also understood that the new indexer/tokenizer can do it in some way when we discussed this issue some time ago.

**Comment by jirikuncar** [comment]
@eamonnmag are you going to target anything in Invenio-Trends?

**Comment by eamonnmag** [comment]
I think the stats are a little outside our use case. But, I've redone the holding pen interface and one of the views could potentially be reused to visualise this info. I wanted to add something similar to what I did for the THOR dashboard (http://dashboard.project-thor.eu/dashboard/data/) where you can select different time periods and look at stats per hour etc. Another example is what is in crossfilter (which is what I use largely - http://square.github.io/crossfilter/)
